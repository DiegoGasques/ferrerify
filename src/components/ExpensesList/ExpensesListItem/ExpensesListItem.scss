.ExpensesListItem {
  margin-bottom: 1.5rem;
  -webkit-box-shadow: 0px 4px 15px -5px rgba(0, 0, 0, 0.5);
  box-shadow: 0px 4px 15px -5px rgba(0, 0, 0, 0.5);
  color: #919090;
}

.ExpensesListItem__header {
  padding: .7rem 1.5rem;
  display: flex;
  justify-content: space-between;
}

.ExpensesListItem__toggle-Btn {
  background: none;
  border: none;
  padding: .2rem .5rem;
  margin: 0;
  text-align: center;
  font-size: 1.8rem;
}

.ExpensesListItem__toggle-Btn:active,
.ExpensesListItem__toggle-Btn:focus
 {
  outline: none;
}

.fa-sort-up {
  transform: translate(0, 4px);
}

.fa-sort-down {
  transform: translate(0, -4px);
}

.ExpensesListItem__body {
  padding: .7rem 1.5rem;
}

.ExpenseListItem__info {
  display: flex;
  justify-content: space-between;
  margin-bottom: 1.5rem;
}

.ExpenseListItem__note {
  margin-bottom: 1.5rem;
}

.ExpenseListItem__amount span,
.ExpenseListItem__note span {
  font-weight: 600;
  color: black;
}

.ExpenseListItem__action-buttons {
  display: flex;
  justify-content: flex-end;
}

.ExpenseListItem__delete-btn,
.ExpenseListItem__edit-btn {
  border: none;
  background: none;
  color: white;
  margin-left: 1.5rem;
  padding: .7rem 1.5rem;
  font-size: 1rem;
  font-weight: bold;
  text-transform: uppercase;
}

.ExpenseListItem__delete-btn {
  background: red;
}

.ExpenseListItem__edit-btn {
  background: green;
}

.ExpenseListItem__edit-btn a{
  text-decoration: none;
  color: inherit;
}