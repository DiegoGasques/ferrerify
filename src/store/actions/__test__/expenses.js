import { addExpense, removeExpense, editExpense } from "../expenses.actions.js";

it("Test for addExpense action creators correct output", () => {
  const newExpense = {
    description: "New expense",
    note: "This is a big expense",
    createdAt: 1000000,
    amount: 15000
  };
  const action = addExpense(newExpense);

  expect(action).toEqual({
    type: "ADD_EXPENSE",
    expense: {
      ...newExpense,
      id: expect.any(String)
    }
  });
});

it("Test for removeExpense action creators correct output", () => {
  const id = "123";
  const action = removeExpense(id);
  expect(action).toEqual({
    type: "REMOVE_EXPENSE",
    id
  });
});

it("Test for editExpense action creators correct output", () => {
  const id = "123";
  const updates = { note: "The updated note" };
  const action = editExpense(id, updates);
  expect(action).toEqual({
    type: "EDIT_EXPENSE",
    id,
    updates
  });
});
