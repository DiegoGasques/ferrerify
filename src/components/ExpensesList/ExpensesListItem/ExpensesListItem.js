import React, { useState } from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import numeral from "numeral";
import moment from "moment";
import { removeExpense } from "../../../store/actions/expenses.actions";

export const ExpensesListItem = ({
  id,
  description,
  createdAt,
  note,
  amount,
  removeExpense
}) => {
  const [isOpen, setIsOpen] = useState(false);
  const handleToggle = () => setIsOpen(!isOpen);
  return (
    <div className="ExpensesListItem">
      <div className="ExpensesListItem__header">
        <h3>{description}</h3>
        <button onClick={handleToggle}>
          {isOpen ? (
            <i class="fas fa-sort-up" />
          ) : (
            <i class="fas fa-sort-down" />
          )}
        </button>
      </div>
      {isOpen && (
        <div className="ExpensesListItem__body active">
          <div className="ExpenseListItem__info">
            <div className="ExpenseListItem__createdAt">
              {moment(createdAt).calendar()}
            </div>
            <div className="ExpenseListItem__amount">
              <span>$</span>
              {numeral(amount).format(0, 0.0)}
            </div>
          </div>
          <div className="ExpenseListItem__note">
            <span>Note: </span>
            {note}
          </div>
          <div className="ExpenseListItem__action-buttons">
            <button
              className="ExpenseListItem__delete-btn"
              onClick={removeExpense}
            >
              delete
            </button>
            <button className="ExpenseListItem__edit-btn">
              <Link to={`/edit/${id}`}>edit</Link>
            </button>
          </div>
        </div>
      )}
    </div>
  );
};

const mapDispatchToProps = (dispatch, { id }) => ({
  removeExpense: () => dispatch(removeExpense(id))
});

export default connect(
  null,
  mapDispatchToProps
)(ExpensesListItem);
