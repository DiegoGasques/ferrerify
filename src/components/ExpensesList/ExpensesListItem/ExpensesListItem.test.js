import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { ExpensesListItem } from "./ExpensesListItem";
import { expensesArr } from "../../../fixtures/expenses";

describe("[COMPONENT] ExpensesListItem", () => {
  it("Tests for handle toggle button being called correctly", () => {
    const wrapper = shallow(<ExpensesListItem {...expensesArr[0]} />);
    expect(wrapper.find(".fa-sort-down")).toBeDefined();
    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper.find(".ExpensesListItem__toggle-Btn").simulate("click");
    expect(wrapper.find(".fa-sort-up")).toBeDefined();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("Tests for removeExpense to have been called on button click", () => {
    const removeExpenseSpy = jest.fn();
    const wrapper = shallow(
      <ExpensesListItem removeExpense={removeExpenseSpy} {...expensesArr[0]} />
    );

    wrapper.find(".ExpensesListItem__toggle-Btn").simulate("click");
    wrapper.find(".ExpenseListItem__delete-btn").simulate("click");
    expect(removeExpenseSpy).toHaveBeenCalled();
  });
});
