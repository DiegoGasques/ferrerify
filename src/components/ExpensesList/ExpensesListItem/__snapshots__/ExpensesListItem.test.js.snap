// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[COMPONENT] ExpensesListItem Tests for handle toggle button being called correctly 1`] = `
<div
  className="ExpensesListItem"
>
  <div
    className="ExpensesListItem__header"
  >
    <h3>
      Rent
    </h3>
    <button
      onClick={[Function]}
    >
      <i
        class="fas fa-sort-down"
      />
    </button>
  </div>
</div>
`;

exports[`[COMPONENT] ExpensesListItem Tests for handle toggle button being called correctly 2`] = `
<div
  className="ExpensesListItem"
>
  <div
    className="ExpensesListItem__header"
  >
    <h3>
      Rent
    </h3>
    <button
      onClick={[Function]}
    >
      <i
        class="fas fa-sort-up"
      />
    </button>
  </div>
  <div
    className="ExpensesListItem__body active"
  >
    <div
      className="ExpenseListItem__info"
    >
      <div
        className="ExpenseListItem__createdAt"
      >
        12/31/1969
      </div>
      <div
        className="ExpenseListItem__amount"
      >
        <span>
          $
        </span>
        1,500
      </div>
    </div>
    <div
      className="ExpenseListItem__note"
    >
      <span>
        Note: 
      </span>
      Monthly payment
    </div>
    <div
      className="ExpenseListItem__action-buttons"
    >
      <button
        className="ExpenseListItem__delete-btn"
      >
        delete
      </button>
      <button
        className="ExpenseListItem__edit-btn"
      >
        <Link
          to="/edit/1"
        >
          edit
        </Link>
      </button>
    </div>
  </div>
</div>
`;
